_MultiTwin_ package for the analysis of multipartite graphs.

_Authors_: Eduardo Corel & Jananan S. Pathmanathan -- 2014-2018.

# Prerequisites

To run the _MultiTwin_ suite, you need the following software. If already installed, you can skip the corresponding paragraph.

|_Software_|_Recommended version_|_Install method_|
|--------|:-----------:|:-----------|
|`Python 3`|version `>= 3.5`, recommended `3.6.5` ; not `Python 2`|`sudo` or your system administrator
|`g++`|version `>= 4.8` |`sudo` or your system administrator
|`igraph`| source `C`  code + `python`  bindings (version `>= 0.7.1` )| `sudo` or your system administrator
|`blast`  | complete NCBI suite (version `>= 2.6` )| _via_ `install.sh`
|`exonerate` | EBI software package (version `>= 2.2` )|_via_ `install.sh`
|`diamond` | Huson lab software package (version `>= 0.9.6`)|_via_ `install.sh`

1. _Python_: Minimum `Python 3.5`, recommended version: `3.6.5`
 Attention: The code will likely NOT work under `Python 2`. 
 * To install the `python-igraph` bindings, as well as the `Tkinter` modules for the graphical interface, the installer script `install.sh` assumes that `pip` and `tk` are available for `python3`. If installation fails, try:

       $ sudo apt-get install python3-pip python3-dev build-essential python3-tk 
2. _C++_: compiler `g++`  version `>= 4.8`
To check what version is currently installed, type:

       $ g++ --version
   If an update is required, type the following commands:

       $ sudo add-apt-repository ppa:ubuntu-toolchain-r/test
       $ sudo apt-get update
       $ sudo apt-get install g++-4.8
       $ sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-4.8 50

3. `igraph` library: 

  * Install the `igraph` C code: download the latest version of the `igraph` library (here `0.7.1`) to your `DOWNLOAD_DIR`

        $ mkdir DOWNLOAD_DIR
        $ cd DOWNLOAD_DIR/.
        $ wget http://igraph.org/nightly/get/c/igraph-0.7.1.tar.gz 
        $ tar -xvf igraph-0.7.1.tar.gz 
        $ rm igraph-0.7.1.tar.gz 
        $ cd igraph-0.7.1
        $ less INSTALL

    and follow the install instructions -- usually

        $ ./configure 
        $ make
        $ sudo make install

    _NB_: You might need to run additionally

        $ sudo ldconfig

       If you get an error message for missing libraries `libxml2-dev` or `zlib1g-dev`, type accordingly

        $ sudo apt-get install libxml2-dev 
        $ sudo apt-get install zlib1g-dev

     and redo the install instructions above.

     _WARNING_: installing the `igraph` library can take a rather long time.
     
  * The `igraph` libraries (_e.g._ `libigraph.so.0`) should be in the directory `/usr/local/lib` and the header files (_e.g._ `igraph.h`) in `/usr/local/include/igraph`. If it is not the case, edit the `BlastProg/makefile` to set the following lines according to your installation

`PATH_IGRAPH_LIB = -L/usr/local/lib`

`PATH_IGRAPH_INC = -I/usr/local/include/igraph`
         
# INSTALLATION of the _MultiTwin_ package

* Download the tar file in a directory `INSTALL_DIR`

       $ cd INSTALL_DIR
       $ tar -xvf MultiTwin1.2.tar
       $ rm MultiTwin1.2.tar
       $ cd MultiTwin1.2/
       $ chmod 751 install.sh
       $ ./install.sh

The executable scripts `cleanblast` and `familydetector` should have been generated in `BlastProg/`. 
If this is not the case, check the version of the `g++` compiler (see *Prerequisites* above): an error should have been generated by the installer.
You will notice that the directories `simtools` and `exonerate` have been generated.  
* Refresh the system variables in the `.bashrc` configuration file, by typing
   
      $ source ~/.bashrc
_NB_: If you need to run the `install.sh` script again, make sure that you edit the `.bashrc` file and remove the lines starting with 
`# These lines have been written by the MULTITWIN install.sh file`
and ending with
`# END of MULTITWIN install section`
to avoid unnecessary cluttering of your `bashrc` file.

* Test if the installation was successful, by running the following:

      $ cd data
      $ bitwin.py -G -a seq.annot -n 30 -b seq.blastp -g genome2seq.txt

The graphical interface with the entered values is displayed; press  the `Run` button.

### The `data`  DIRECTORY initially contains:

| File | DESCRIPTION | COMMENT |
|--------|:-----------|:------|
|`seq.blastp` | The result of the `BLAST` or `diamond`  run.| First compulsory argument for the `bitwin.py`  program.|
|`genome2seq.txt` | Contains the correspondence between sequences and genomes: `X TAB Y`  where X is the identifier of a genome and Y the identifier of a sequence. | Second compulsory argument for `bitwin.py` .| 
|`seq.annot` | Annotation file for both sequences and genomes. Fields having no match for a given row are shown with a dash ("-"). | This file is optional (option `-a` ).|

### The previous command produces the following hierarchy of directories and files:
1. in `data` 

  | File | DESCRIPTION | COMMENT |
  |--------|:-----------|:------|
  |`seq.blastp.cleanNetwork` |  Sequence-similarity graph obtained by keeping the best reciprocal hit (`>=30%` ID) with reciprocal minimum coverage of `80%` | Sequence IDs are renumbered.|
  |`seq.blastp.cleanNetwork.genes` |  List of new sequence IDs | Internal use only.|
  | `seq.blastp.cleanNetwork.dico` |   Correspondence between initial IDs and sequence ids. | Internal use only.|

2. in `data/graphs30/` 

  | File | DESCRIPTION | FORMAT |
  |--------|:-----------|:------|
  |`CC.nodes` | Community file for gene families (at 30% identity)| `node TAB communityID`| 
	  222147474	17
	  153932043	12
	  117618069	15
	  187476667	23
	  59712147	6
	  (...)

  ||||
  |--------|:-----------|:------|
  | `CC.edges` | Fasta-formatted list of edges for gene families (includes singletons _i.e._ families without edges).| `head TAB tail`|
    >CC1
	>CC5
	7	13
	>CC6
	8	11
	(...)

  ||||
  |--------|:-----------|:------|
  |`CC.info` | Graph statistics for the gene families.|`CCid TAB nbNodes TAB nbEdges`|

    #CCid	  nodes	  edges
	1	1	0
	2	1	0
	3	1	0
	(...)

  ||||
  |--------|:-----------|:------|
  |`graph.edges` | The bipartite graph obtained by factoring the `GenomeToSeq.txt`  ROOT graph by the gene family file| The weight in the third column is 1+_log_(nb of original edges).|
        
    38	1	1.00
	28	22	1.00
	38	20	1.00
	(...)

  ||||
  |--------|:-----------|:------|
  |`graph.trail` |The corresponding trail file.|`UniqID TAB currentID`|
    # factorgraph.py -d graphs30 -k UniqID -c graphs30/CC.nodes -s '\t' genome2seq.txt graph.edges graph.trail 
    # Root directory: DATA_DIRECTORY`  
    93006436	6 
    213155360	39
    (...)

  ||||
  |--------|:-----------|:------|
  |`graph0.edges` |The bipartite graph obtained by removing all degree one gene families.||
    28	22	1.00
    29	23	1.00 
    40	6	1.69 
    (...)

  ||||
  |--------|:-----------|:------|
  |`graph0.twins` |The community file describing the twin mapping. |`currentID TAB twinID`|
    15	1
	14	3
	17	4
	(...)

  ||||
  |--------|:-----------|:------|
  |`graph0.twin_comp` |The community file describing the twin+support overlapping clustering.| _idem_| 
    15	1
	14	3
	17	4
	23	0
	(...)

3. in `data/graphs30/TwinQuotient/` 

  ||||
  |--------|:-----------|:------|
  |`graph1.edges` |The bipartite graph obtained by factoring `graph0.edges`  by the `graph0.twins`  file.||
    17	5	1.00
	16	0	1.00
	10	6	1.00
	(...)

  ||||
  |--------|:-----------|:------|
  |`graph1.trail` |The corresponding trail file.|`UniqID TAB currentID`|
    # factorgraph.py -d 'TwinQuotient' -k 'UniqID' -c 'graph0.twins' -s '\t' -t 'graph.trail' graph0.edges graph1.edges graph1.trail
	# Root directory: DATA_DIRECTORY
	59251	10
	222147474	4
	(...)

||||
  |--------|:-----------|:------|
  |`graph1.bic_comp` |List of biconnected components| `node TAB component`|
    16	2
    4	0
    4	3
	5
	(...)

  ||||
  |--------|:-----------|:------|
  |`graph1.art` |Articulation points of `graph1` and biconnected components they belong to (comma-separated)|
    5	4,5,6,7
    12	7,8,12
    4	0,3
	(...)


  ||||
  |--------|:-----------|:------|
  |`graph1.twins` | Twin file of `graph1`  ||
    1	0
	0	3
	3	6
	(...)

  ||||
  |--------|:-----------|:------|
  |`graph1.twin_comp` |   Community file describing the twin support overlapping clustering of `graph1`.||
    1	0
	0	3
	(...)

  ||||
  |--------|:-----------|:------|
  |`graph1.config` |  XML Configuration file for the `description.py`   script (embedded in `bitwin.py`)||


  ||||
  |--------|:-----------|:------|
  |`graph1.desc` | Description file: every `Module` is a twin of gene families plus its support|The fields at the beginning of each row match the column names in the annotation file `seq.annot`).
	Module 2
    CC      {'6402': 5}
    DESCRIPTION     {'Outer membrane receptor proteins, mostly Fe transport': 5}
    Genome Project ID       {'PRJNA61563': 1, 'PRJNA59083': 1, 'PRJNA52575': 1, 'PRJNA59899': 1}
    COGID   {'COG1629': 5}
    COGCAT  {'P': 5}
    SPECIES {'59899': 1, '61563': 1, '59083': 2, '52575': 1}
    Species {'Bordetella avium 197N': 1, 'Actinobacillus pleuropneumoniae serovar 11 str. 56153': 1, 'Acinetobacter baumannii AB0057': 1, 'Achromobacter xylosoxidans A8': 1}
    Shorthand ID    {'Bav': 1, 'Aba': 1, 'Apl': 1, 'Axy': 1}
    Pathogen        {'2': 1, '1': 3}
    PATH    {'2': 1, '1': 4}
	(...)

  ||||
  |--------|:-----------|:------|
  |`graph1.xml_desc` | XML version of the description file (easier to parse than `graph1.desc`  | The same `Module 2`  is described as follows| 

	<mod name="Module" id="2">
    <content>
        <attr name="CC">
                {'6402': 5}
        </attr>
        <attr name="DESCRIPTION">
                {'Outer membrane receptor proteins, mostly Fe transport': 5}
        </attr>
        <attr name="Genome Project ID">
                {'PRJNA61563': 1, 'PRJNA59083': 1, 'PRJNA52575': 1, 'PRJNA59899': 1}
        </attr>
        <attr name="COGID">
                {'COG1629': 5}
        </attr>
        <attr name="COGCAT">
                {'P': 5}
        </attr>
        <attr name="SPECIES">
                {'59899': 1, '61563': 1, '59083': 2, '52575': 1}
        </attr>
        <attr name="Species">
                {'Bordetella avium 197N': 1, 'Actinobacillus pleuropneumoniae serovar 11 str. 56153': 1, 'Acinetobacter baumannii AB0057': 1, 'Achromobacter xylosoxidans A8': 1}
        </attr>
        <attr name="Shorthand ID">
                {'Bav': 1, 'Aba': 1, 'Apl': 1, 'Axy': 1}
        </attr>
        <attr name="Pathogen">
                {'2': 1, '1': 3}
        </attr>
        <attr name="PATH">
                {'2': 1, '1': 4}
        </attr>
    </content>


For more information on each type of file, see the `README_data`  file with the complete attached data `MultiTwin1.2_Data.tar.gz` 

